NanoProd/plugins/TauSpinnerTableProducer.cc

```diff
@@ -0,0 +1,216 @@
+  for (const auto &daughterRef : boson.daughterRefVector()) {
+    if (std::abs(daughterRef->pdgId()) == 15)
+      taus.push_back(getLastCopy(daughterRef));
+  }
+}
+
+void TauSpinnerTableProducer::getTauDaughters(reco::GenParticleRefVector &tau_daughters, const reco::GenParticle &tau) {
+  for (auto daughterRef : tau.daughterRefVector()) {
+    if (tauProdsIds.find(std::abs(daughterRef->pdgId())) == tauProdsIds.end())
+      getTauDaughters(tau_daughters, *daughterRef);
+    //throw std::runtime_error("Unknown tau decay product, pdgId = " + std::to_string(daughterRef->pdgId()));//FIXME
+    else
+      tau_daughters.push_back(daughterRef);
+  }
+}
+
+void TauSpinnerTableProducer::beginJob() {
+  // Initialize TauSpinner
+  Tauolapp::Tauola::setNewCurrents(0);
+  Tauolapp::Tauola::initialize();
+  LHAPDF::initPDFSetByName(tauSpinnerPDF_);
+  TauSpinner::initialize_spinner(ipp_, ipol_, nonSM2_, nonSMN_, cmsE_);
+}
+
+void TauSpinnerTableProducer::produce(edm::Event &event, const edm::EventSetup &setup) {
+  // Input gen-particles collection
+  auto const &genParts = event.get(genPartsToken_);
+
+  // Output table
+  auto wtTable = std::make_unique<nanoaod::FlatTable>(1, branch_, true);
+  wtTable->setDoc("TauSpinner weights");
+
+  // Search for boson
+  edm::RefVector<edm::View<reco::GenParticle>> bosons;
+  getBosons(bosons, genParts);
+  if (bosons.size() != 1) {  //no boson found or more than one found, produce empty table (expected for non HTT sample)
+    event.put(std::move(wtTable));
+    return;
+  }
+
+  // Search for taus from boson decay
+  reco::GenParticleRefVector taus;
+  getTaus(taus, *bosons[0]);
+  if (taus.size() != 2) {  //boson does not decay to tau pair, produce empty table (expected for non HTT sample)
+    event.put(std::move(wtTable));
+    return;
+  }
+
+  // Get tau daughters and convert all particles to TauSpinner format
+  TauSpinner::SimpleParticle simple_boson = convertToSimplePart(*bosons[0]);
+  std::array<TauSpinner::SimpleParticle, 2> simple_taus;
+  std::array<std::vector<TauSpinner::SimpleParticle>, 2> simple_tau_daughters;
+  for (size_t tau_idx = 0; tau_idx < 2; ++tau_idx) {
+    simple_taus[tau_idx] = convertToSimplePart(*taus[tau_idx]);
+    reco::GenParticleRefVector tau_daughters;
+    getTauDaughters(tau_daughters, *taus[tau_idx]);
+    for (const auto &daughterRef : tau_daughters)
+      simple_tau_daughters[tau_idx].push_back(convertToSimplePart(*daughterRef));
+  }
+
+  // Compute TauSpinner weights and fill table
+  std::array<double, 2> weights;
+  for (const auto &theta : theta_vec_) {
+    // Can make this more general by having boson pdgid as input or have option for set boson type
+    TauSpinner::setHiggsParametersTR(-cos(2 * M_PI * theta.second),
+                                     cos(2 * M_PI * theta.second),
+                                     -sin(2 * M_PI * theta.second),
+                                     -sin(2 * M_PI * theta.second));
```

Comment:
Is -sin twice intended? or one should be +sin?
```suggestion
    const double angle = 2 * M_PI * theta.second;
    const double cos_theta = std::cos(angle);
    const double sin_theta = std::sin(angle);
    TauSpinner::setHiggsParametersTR(-cos_theta, cos_theta, -sin_theta, -sin_theta);
```

Reply:
It is correct, cf. https://arxiv.org/pdf/1802.05459, p. 21